### PR TITLE
subscriptions to notifications

### PR DESCRIPTION
**Pull request by konstan**
This PR adds the main left vertical tab Notifications with an ability to create and manage notification subscriptions and their alert methods.

This is an evolution of https://github.com/nuvla/ui/pull/492 which was done for infra services only in a pre-canned way and is discarded in the favor of this one.

### CODE REVIEW

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 1]
```diff
@@ -0,0 +1,930 @@
...
+         ]))))
+
+
+(defn TabMethods
+  []
+  (let [tr (subscribe [::i18n-subs/tr])
+        all-methods (subscribe [::subs/notification-methods])]
+    (dispatch [::events/get-notification-methods])
+    (fn []
+      (let []
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:876]: Remove empty let

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 2]
```diff
@@ -0,0 +1,930 @@
...
+        [ui/TabPane
+         [MenuBarNotificationMethod]
+         [:div methods]
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:879]: Wrong methods var is used

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 3]
```diff
@@ -0,0 +1,930 @@
...
+    [taoensso.timbre :as timbre]))
+
+
+(defn save-callback-notification-subscription-config
+  []
+  (dispatch-sync [::events/set-validate-form? true])
+  (dispatch-sync [::events/validate-notification-subscription-config-form])
+  (let [form-valid? (get @re-frame.db/app-db ::spec/form-valid?)]
+    (when form-valid?
+      (do
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:29]: `when` is able to have multiple statement in it's body. The do is useless

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 4]
```diff
@@ -0,0 +1,930 @@
...
+                :on-click #(dispatch [::events/open-add-update-notification-method-modal {} true])}]])
+
+
+(defn subs-notif-method-dropdown
+  [current-value notif-methods save? collection subs-conf-id]
+  ^{:key current-value}
+  [ui/FormDropdown
+   {:selection     true
+    :fluid         false
+    :default-value (if (not (nil? current-value))
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:110]: not nil? use some?

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 5]
```diff
@@ -0,0 +1,930 @@
...
+
+(defn subs-status-dropdown
+  [current-value]
+  ^{:key current-value}
+  [ui/Dropdown
+   {:selection     true
+    :fluid         false
+    :default-value current-value
+    :on-change     (ui-callback/value
+                     #(do (dispatch-sync [::events/update-subscription :status %])))
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:234]: try a maximum to not use dispatch-sync. do is useless

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 6]
```diff
@@ -0,0 +1,930 @@
...
+          [uix/Button {:text     (@tr [:save])
+                       :positive true
+                       :disabled true  #_(when-not @form-valid? true)
+                       :active   true
+                       :on-click #(save-callback-subscription ::spec/subscription)}]]]))))
+
+(defn select-resources
+  []
+  (let [component-type (subscribe [::subs/collection])]
+    [
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:294]: 2 [[ useless

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 7]
```diff
@@ -0,0 +1,930 @@
...
+           [uix/TableRowField "category", :editable? false, :default-value category, :required? false]
+           [uix/TableRowField "resource"
+            :editable? false
+            :default-value [history/link (str "api/" resource-id) resource-id]
+            :required? false]]]
+
+         [ui/ModalActions
+          [uix/Button {:text     (@tr [:save])
+                       :positive true
+                       :disabled true  #_(when-not @form-valid? true)
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:287]: intended?

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 8]
```diff
@@ -0,0 +1,930 @@
...
+                    (dispatch [::events/update-notification-subscription-config name-kw value])
+                    (dispatch [::events/validate-notification-subscription-config-form]))
+        notif-methods (subscribe [::subs/notification-methods])
+        collection (subscribe [::subs/collection])
+        criteria-metric (subscribe [::subs/criteria-metric])
+        components-number (subscribe [::subs/components-number])]
+    (dispatch [::events/get-notification-methods])
+    (dispatch [::events/set-components-number 0])
+    (dispatch [::events/reset-tags-available])
+    #_(dispatch [::events/update-notification-subscription-config :resource-filter ""])
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:376]: intended?

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 9]
```diff
@@ -0,0 +1,930 @@
...
+        visible? (subscribe [::subs/edit-subscription-modal-visible?])
+        validate-form? (subscribe [::subs/validate-form?])
+        form-valid? (subscribe [::subs/form-valid?])
+        subscription (subscribe [::subs/subscription])
+        on-change (fn [name-kw value]
+                    (dispatch [::events/update-subscription name-kw value])
+                    (dispatch [::events/validate-subscription-form]))]
+    (dispatch [::events/get-notification-methods])
+    (fn []
+      (let [editable? true
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:251]: editable? not used

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 10]
```diff
@@ -0,0 +1,930 @@
...
+    :options       (map (fn [v] {:key v, :value v, :text v}) ["enabled" "disabled"])}])
+
+
+(defn edit-subscription-modal
+  []
+  (let [tr (subscribe [::i18n-subs/tr])
+        notif-methods (subscribe [::subs/notification-methods])
+        visible? (subscribe [::subs/edit-subscription-modal-visible?])
+        validate-form? (subscribe [::subs/validate-form?])
+        form-valid? (subscribe [::subs/form-valid?])
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:244]: not used

File: code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+(reg-event-db
+  ::close-add-update-notification-method-modal
+  (fn [db _]
+    (assoc db ::spec/notification-method-modal-visible? false)))
+
+
+(reg-event-fx
+  ::delete-notification-method
+  (fn [{:keys [db]} [_ id]]
+    {:db                  db
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs:61]: :db db is useless
You can also replace {:keys [db]} by _

File: code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+
+(reg-event-db
+  ::set-notification-methods
+  (fn [db [_ notification-methods]]
+    (assoc db ::spec/notification-methods notification-methods)))
+
+
+(reg-event-fx
+  ::get-notification-methods
+  (fn [{:keys [db]} [_]]
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs:36]: [_] should be replaced by _
Like that clojurescript will do the destructuring with first

File: code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+  (fn [db [_]]
+    (assoc db ::spec/notification-subscription-config {})
+    (assoc db ::spec/components-number 0)
+    (assoc db ::spec/resource-tags-available [])
+    (assoc db ::spec/collection nil)))
+
+
+(reg-event-fx
+  ::toggle-enabled
+  (fn [{:keys [db]} [_ rid enabled]]
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs:242]: db not used replace with _

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 14]
```diff
@@ -0,0 +1,930 @@
...
+                            "nuvlabox"               nuvlabox-subs-confs}]
+        [ui/TabPane
+         [MenuBarSubscription]
+         (if (empty? @subscription-configs)
+           [ui/Message (str/capitalize (@tr [:no-subscription-configs-defined]))]
+           (for [[resource-kind resource-subs-confs] subs-confs-all]
+             (if-not (empty? resource-subs-confs)
+               ^{:key resource-kind}
+               [uix/Accordion
+                [:<>
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:802]: :<> and :div useless

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 15]
```diff
@@ -0,0 +1,930 @@
...
+  []
+  (let [tr (subscribe [::i18n-subs/tr])
+        notif-methods (subscribe [::subs/notification-methods])
+        subscriptions (subscribe [::subs/notification-subscriptions])
+        visible? (subscribe [::subs/notification-subscriptions-modal-visible?])]
+    (dispatch [::events/get-notification-methods])
+    (dispatch [::events/get-notification-subscriptions])
+    (fn []
+      (let [header (@tr [:subscriptions-manage])]
+        [:div]
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:187]: useless

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs [hunk 16]
```diff
@@ -0,0 +1,930 @@
...
+               [ui/TableCell
+                [ui/Input
+                 {:type      "text"
+                  :name      "Value"
+                  :read-only false
+                  :value     (:value criteria)
+                  :on-change (ui-callback/value #(on-change :criteria {:value %}))}]]]
+              )]]
+          [ui/Header {:as "h3"} "Notification"]
+          [:<>
```
Comment (0xbase12) [inline code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:595]: :<> useless

**Review by 0xbase12 — CHANGES_REQUESTED** [review]

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 1 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:876]: done

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 2 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:879]: removed

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 3 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:29]: changed

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 4 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:110]: done

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 5 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:234]: done

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 6 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:294]: done

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 7 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:287]: removed

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 8 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:376]: removed

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 9 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:251]: removed

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 10 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:244]: removed

File: code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs (hunk 11 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs:61]: changed

File: code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs (hunk 12 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs:36]: changed

File: code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs (hunk 13 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/events.cljs:242]: replaced

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 14 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs]: done

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 15 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:181]: removed

File: code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs (hunk 16 above)
Comment (konstan) [reply code/src/cljs/sixsq/nuvla/ui/notifications/views.cljs:585]: removed

**Comment by konstan** [comment]
@0xbase12 I responded to your comments. Please check.

**Comment by 0xbase12** [comment]
Thanks. One more thing
[
Screenshot 2021-02-16 at 09 27 42
](url)

**Comment by konstan** [comment]
checking

**Comment by konstan** [comment]
@0xbase12 fixed.

**Comment by 0xbase12** [comment]
Still having an error message and validation is failing at open of the modal.
Screenshot 2021-02-16 at 09 46 47

**Comment by konstan** [comment]
@0xbase12 updated. Is it better?

**Comment by 0xbase12** [comment]
Yes, the message is gone from the console. But for modal "new notification method", the validation of form is failing at open

**Comment by konstan** [comment]
@0xbase12 pushed a change - please  check.

**Comment by 0xbase12** [comment]
Working

**Pull request merged**